Added support for logging via the supervisor

ssttevee opened the pull request:
Resolves #13. 

The batching logic mirrors that of the logger from the worker.js retrieved by following the instructions from #16.

The logger may be used directly:
```
nodego.InfoLogger.Println("Hello World!")
```
or via the `log` package after calling `nodego.OverrideLogger()`
```
func init() {
	nodego.OverrideLogger()
}

...

log.Println("Hello World")
```

ssttevee commented:
I signed it!

iangudger commented on examples/logging.go:
```diff
@@ -0,0 +1,22 @@
+	"net/http"
+
+	"../nodego"
+	"log"
```
Put this import with the other stdlib imports.

iangudger commented on nodego/env.go:
```diff
@@ -0,0 +1,43 @@
+	_FUNCTION_TRIGGER_TYPE    = os.Getenv("FUNCTION_TRIGGER_TYPE")
+	_FUNCTION_NAME            = os.Getenv("FUNCTION_NAME")
+	_FUNCTION_TIMEOUT_SEC, _  = strconv.ParseInt(os.Getenv("FUNCTION_TIMEOUT_SEC"), 10, 64)
+	// HTTP header field that is added to Worker response to signalize problems with
```
Comments documenting declarations should begin with the name of the thing being described and end in a period.
https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

ssttevee replied:
Should these comments be left out and replaced with a reference to the `worker.js` file?

iangudger commented on nodego/env.go:
```diff
@@ -0,0 +1,43 @@
+)
+
+var (
+	_CODE_LOCATION_DIR        = os.Getenv("CODE_LOCATION")
```
Please use mixed caps.
https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,162 @@
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
```
Please use the std lib context.

iangudger commented on examples/logging.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+func init() {
+	nodego.OverrideLogger()
```
Almost everything in GCF is request based, including logging.  GCF inserts middleware in the request path to report logs on a per-request basis, so I don't see why we shouldn't do the same. GCF guarantees that your instance will only receive a single concurrent request to make this possible.

Ex:
```
func LoggingMiddleware(h http.HandlerFunc ) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		startLogging()
		h(w, r)
		endLogging()
	}
}

func main() {
	flag.Parse()
	http.HandleFunc(nodego.HTTPTrigger, LoggingMiddleware(func(w http.ResponseWriter, r *http.Request) {
		log.Println("This is a log message from Go!")
	}))

	nodego.TakeOver()
}
```

ssttevee replied:
How should the cleanup work? It doesn't look like it's possible to get the existing writer from the std logger.

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,162 @@
+	severity string
+}
+
+func (w *supervisorWriter) Write(p []byte) (n int, err error) {
```
It seems like we aught to be able to detect if we are running locally and log to STDERR.

ssttevee commented:
I added a `WithLogger` middleware function that gets the `Function-Execution-Id` header and includes in new log entries. I'm not sure there's anything else for it to do.

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,189 @@
+	return func(w http.ResponseWriter, r *http.Request) {
+		functionExecutionId = r.Header.Get("Function-Execution-Id")
+		handler.ServeHTTP(w, r)
+	}
```
If I recall correctly, worker.js sends all pending log entries when the request returns. This avoids having log lines buffered too long and avoids losing log lines if the instance is killed shortly after handling a request.

ssttevee replied:
I feel like this should not be too much of a problem unless there's more than a few full batches of logs since the worker should be sending off log batches as fast as possible.

iangudger replied:
Here are some scenarios to consider:
1. You are testing your app. You make a request and want to see the logs. You check the logs, but none show up. You wait a few minutes, refresh again and still don't see the logs.
1. You have a low-traffic app that gets one request every few hours. You notice that you rarely get logs from it and when you do they are greatly delayed (and possibly truncated).
1. Similar to the first scenario, only in this case you only see some of the log lines instead of none of them.

In each case, the user is likely assume something is broken (most likely the unofficial Go runtime, but possibly Stackdriver Logging).

iangudger replied:
Remember that the post-request flush does not need to be synchronous, kicking off an asynchronous task is probably best.

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,189 @@
+	return len(entry.TextPayload), nil
+}
+
+func logBatchReportWorker() {
```
Optional: It might be nice to have a ticker or something that adds a time trigger on log buffering in addition to the number/length based trigger you currently have. Maybe add a TODO?

ssttevee commented:
I have moved things around and un-exported the functions that don't need to be exported. I also added the flushing mechanism but somewhere in the back of my mind, it doesn't feel like the best approach.

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,255 @@
+}
+
+// WithLoggerFunc is the same as WithLogger but accepts a handler function.
+func WithLoggerFunc(handler func(w http.ResponseWriter, r *http.Request)) http.HandlerFunc {
```
You should be able to just accept a http.HandlerFunc and then not need to cast it below.

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,255 @@
+		loggingContext.Unlock()
+
+		defer func() {
+			go flushLogBatches()
```
Ideally all log lines tagged with a specific execution ID should be flushed. This can be guaranteed by moving this line below the resetting of loggingContext.functionExecutionID.

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,255 @@
+		loggingContext.Unlock()
+
+		defer func() {
+			go flushLogBatches()
```
Why not just call startNewLogBatch? There is already a goroutine waiting that will flush as soon as it can. Plus, that function is already called synchronously in handler execution elsewhere.

ssttevee replied:
This is what I was trying to say before. In fact, there doesn't even need to be a `startNewLogBatch` call because the goroutine will create a new batch if the last one is not empty. `flushLogBatches` actually sends all the remaining log batches asyncronously.

iangudger replied:
`startNewLogBatch` will kick off the flush the next time the `logBatchReportWorker` goroutine is scheduled. Starting a new goroutine won't help as there is no guarantee that it will be scheduled before the already waiting `logBatchReportWorker` goroutine.

ssttevee replied:
If there is an empty batch at the end of the queue, then calling `startNewLogBatch` will result in two empty batches at the end. When `logBatchReportWorker` finds an empty batch, it will wait until something is added to it, and since new logs are only added to the last batch, it will be waiting forever.

`flushLogBatches` will go through all remaining batches regardless of whether they are empty are not and finally creating a new empty batch when the queue is dry.

There does not need to be a guarentee that it runs before a `logBatchReportWorker` loop. If it runs, afterwards and there are no more batches, it will just won't do anything. If it runs before and takes the last batch, `logBatchReportWorker` will wait until `flushLogBatches` creates a new batch and will just wait until that new batch has log added to it.

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,255 @@
+}
+
+func killInstance() {
+	postToSupervisor("/_ah/kill", nil, supervisorKillTimeout)
```
Log error?

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,255 @@
+	}
+}
+
+// WithLogger returns an http handler that reads the function execution id,
```
http.Handler or HTTP handler
execution ID

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,255 @@
+
+const isoTimeFormat = "2006-01-02T15:04:05.999Z07:00"
+
+func (w supervisorWriter) Write(p []byte) (n int, err error) {
```
Could you add the comment:
Write implements io.Writer.Write.

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,255 @@
+	batch.payloadLength += len(entry.TextPayload)
+}
+
+var loggingContext struct {
```
Optional: What you have now is fine, but what I was envisioning was a named type so the following can happen:
logBatchReportWorker becomes a method on the new named type.
startNewLogBatch becomes a method on the new named type.
loggerMiddleware calls a setFunctionExecutionID method on the new named type instead of reaching into its internals.
The locked logic in supervisorWriter.Write moves to a method on the new named type. All of that logic is more tied into the log management than actually creating the log data.

iangudger commented:
Also update the README.md file?

It still says "Logging is not supported (yet)."
https://github.com/GoogleCloudPlatform/cloud-functions-go#limitations

iangudger commented:
All of the new files need a copyright/license header at the top. Ex:
https://github.com/GoogleCloudPlatform/cloud-functions-go/blob/master/nodego/nodego.go

ssttevee commented:
Refactored functions related the mutex into an encapsulated named type as per your suggestion. I also fixed some other bugs here and there.

ssttevee commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,255 @@
+		return err
+	}
+
+	resp, err := ctxhttp.Do(ctx, nil, req)
```
I previously had not used it, but codeclimate was complaining about cognitive complexity.

iangudger replied:
You can ignore codeclimate

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,315 @@
+
+		defer func() {
+			loggingCtx.setExecutionID("")
+			go loggingCtx.flush()
```
Ok, The current synchronization seems correct, although it could probably be simplified. That does not need to happen in this PR though.

* I agree that the current flush can be eliminated. It does not seem to serve a useful purpose.
* I agree that starting new log batches is not needed to ensure that logs are flushed in a timely manor.

That said, I would still like to see new log batches created in the middleware. Here is why:
The new execution ID is only picked up when starting a log batch. As it currently stands, if there is a currently going log batch from before the handler starts, it could get either no execution ID attached or even the wrong execution ID attached.

ssttevee replied:
The log batches themselves has no reference to any execution ID. Execution IDs are attached to the log entry when they are created.

That being said, I do understand wanting to separate batches by execution.

igudger replied:
Oh, good point. Well, in that case I guess we can just remove the flush.

iangudger commented on nodego/supervisor.go:
```diff
@@ -0,0 +1,315 @@
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+	}
+	// Exit code 16 is copied over from worker.js
```
One last style thing: even these comments ending in worker.js should end in a period.

iangudger commented on nodego/env.go:
```diff
@@ -1,3 +1,58 @@
+	"time"
+)
+
+// Variables copied from worker.js
```
This one and the one below too please.
